cmd/src/debug_common.go

```diff
@@ -0,0 +1,470 @@
+package main
+
+import (
+	"archive/zip"
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+	"sync"
+	"syscall"
+
+	"github.com/cockroachdb/errors"
+	"github.com/sourcegraph/src-cli/internal/exec"
+	"golang.org/x/sync/semaphore"
+)
+
+/*
+General Stuff
+TODO: file issue on the existence of OAuth signKey which needs to be redacted
+TODO: Create getSiteConfig function
+*/
+
+type archiveFile struct {
+	name string
+	data []byte
+	err  error
+}
+
+// setOpenFileLimits increases the limit of open files to the given number. This is needed
+// when doings lots of concurrent network requests which establish open sockets.
+func setOpenFileLimits(n uint64) error {
+
+	var rlimit syscall.Rlimit
+	err := syscall.Getrlimit(syscall.RLIMIT_NOFILE, &rlimit)
+	if err != nil {
+		return err
+	}
+
+	rlimit.Max = n
+	rlimit.Cur = n
+
+	return syscall.Setrlimit(syscall.RLIMIT_NOFILE, &rlimit)
+}
+
+// setupDebug takes the name of a base directory and returns the file pipe, zip writer,
+// and context needed for later archive functions. Don't forget to defer close on these
+// after calling setupDebug!
+func setupDebug(base string) (*os.File, *zip.Writer, context.Context, error) {
+	// open pipe to output file
+	out, err := os.OpenFile(base, os.O_CREATE|os.O_RDWR|os.O_EXCL, 0666)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to open file: %w", err)
+	}
+	// increase limit of open files
+	err = setOpenFileLimits(64000)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to set open file limits: %w", err)
+	}
+	// init zip writer
+	zw := zip.NewWriter(out)
+	// init context
+	ctx := context.Background()
+
+	return out, zw, ctx, err
+}
+
+// TODO: Currently external services and site configs are pulled using the src endpoints
+
+// getExternalServicesConfig calls src extsvc list with the format flag -f,
+//and then returns an archiveFile to be consumed
+func getExternalServicesConfig(ctx context.Context, baseDir string) *archiveFile {
+	const fmtStr = `{{range .Nodes}}{{.id}} | {{.kind}} | {{.displayName}}{{"\n"}}{{.config}}{{"\n---\n"}}{{end}}`
+
+	f := &archiveFile{name: baseDir + "/config/external_services.txt"}
+	f.data, f.err = exec.CommandContext(ctx, os.Args[0], "extsvc", "list", "-f", fmtStr).CombinedOutput()
+
+	return f
```

Comment:
```suggestion
	return archiveFileFromCommand(ctx, "/config/external_services.txt",  os.Args[0], "extvcs", "list", "-f", "fmtStr)
```

(1) I'd suggest writing the following function so that the call above is simplified.
(2) I'd suggest replacing all other such uses with a similar call to avoid repetition
(3) I'd modify the following functions so that the error context contains the cmd/args as well as the stdout/stderr of the command (what we added yesterday in [these lines](https://github.com/sourcegraph/src-cli/pull/731/files#diff-497c6ab884ea56cd9de420bc0819ba07bba05fa0288c5ac93d521b4a1b5608a6R328-R330)).

```go
func archiveFileFromCommand(ctx context.Context, path, cmd string, args ...string) *archiveFile {
	f := &archiveFile{name: baseDir + "/config/external_services.txt"}
	f.data, f.err = exec.CommandContext(ctx, cmd, args...).CombinedOutput()
	return f
}
```

Reply:
@efritz Hows this?

```golang
func archiveFileFromCommand(ctx context.Context, baseDir, path, cmd string, args ...string) *archiveFile {
	f := &archiveFile{name: filepath.Join(baseDir, path)}
	f.data, f.err = exec.CommandContext(ctx, cmd, args...).CombinedOutput()
	if f.err != nil {
		f.err = errors.Wrapf(f.err, "executing command: %s %s: received error: %s", cmd, strings.Join(args, " "), f.data)
	}
	return f
}
```
```golang
func getExternalServicesConfig(ctx context.Context, baseDir string) *archiveFile {
	const fmtStr = `{{range .Nodes}}{{.id}} | {{.kind}} | {{.displayName}}{{"\n"}}{{.config}}{{"\n---\n"}}{{end}}`
	return archiveFileFromCommand(ctx, baseDir, "/config/external_services.txt", os.Args[0], "extsvc", "list", "-f", fmtStr)
}
```
It seems a little harder to read? This feels like a stepping stone to the refactor of the archiveX functions. Something like running over [][]string{[path, cmd]}

Reply:
This use one is the least attractive to me --
```golang
// get kubectl logs for pod containers
func getPodLog(ctx context.Context, podName, containerName, namespace, baseDir string) *archiveFile {
	return archiveFileFromCommand(ctx, baseDir, path.Join("/kubectl/pods/", podName, "/", containerName, ".log"), "kubectl", "-n", namespace, "logs", podName, "-c", containerName)
}

//func getPodLog(ctx context.Context, podName, containerName, namespace, baseDir string) *archiveFile {
//	f := &archiveFile{name: baseDir + "/kubectl/pods/" + podName + "/" + containerName + ".log"}
//	f.data, f.err = exec.CommandContext(ctx, "kubectl", "-n", namespace, "logs", podName, "-c", containerName).CombinedOutput()
//	return f
//}
```

Reply:
Sorry, no reason to separate baseDir and path in `archiveFileFromCommand`. This should simplify it a bit:

```go
func archiveFileFromCommand(ctx context.Context, path, cmd string, args ...string) *archiveFile {
        f := &archiveFile{name: path}
        f.data, f.err = exec.CommandContext(ctx, cmd, args...).CombinedOutput()
        if f.err != nil {
                f.err = errors.Wrapf(f.err, "failed to run: `%s %s`: %s", cmd, strings.Join(args, " "), f.data)
        }
        return f
}
```

Then the `getPodLog` becomes this, which is simpler:

```go
func getPodLog(ctx context.Context, podName, containerName, namespace, baseDir string) *archiveFile {
        path := path.Join(baseDir, "/kubectl/pods/", podName, "/", containerName)+".log"
        return archiveFileFromCommand(ctx, path, "kubectl", "logs", podName, "-n", namespace, "-c", containerName)
}
```

Or you could format it a bit differently to better emphasize the literal command (kubectl and on) from the execution details.

```go
func getPodLog(ctx context.Context, podName, containerName, namespace, baseDir string) *archiveFile {
        return archiveFileFromCommand(
                ctx, 
                path.Join(baseDir, "/kubectl/pods/", podName, "/", containerName)+".log", 
                "kubectl", "logs", podName, 
                "-n", namespace, 
                "-c", containerName,
        )
}
```

Reply:
Oh I like that refactor a lot! Will revise

Reply:
Very pleased with how this refactor turned out